### DISCUSSION TITLE
C++ 11 Adoption

### DISCUSSION THREAD

**Discussion by LeStarch** [nasa/fprime#777] [Q&A] [answered]
Community: is there any concern if we were to update to C++11 standards?  This would require compiler support for C++11.  We'd like to know if anyone out there would be unable to work with this change.

Thanks.

**Comment by astroesteban** [comment]
Why not skip C++11 and jump to C++14? C++14 fixes several holes in C++11. It's basically a hotfix for 11.

### Language Features

- C++11’s version of constexpr implies const for member functions, this is changed in C++14
- C++11 is missing auto return type deduction for regular functions (lambdas have it)
- C++11 does not have auto or variadic lambda parameters
- C++14 adds [[deprecated]] attribute
- C++14 adds ' digit separator, example: 1'000'000
- `constexpr` functions can be more than just a single return statement in C++14

### Library Features

- `std::make_unique` was added in C++14, which enables the “no raw new” standard
- C++11 doesn’t have std::exchange
- C++14 adds some constexpr support for std::array cbegin, cend, crbegin and crend free functions added for consistency with begin and end free functions and member functions added to standard containers in C++11.”

**Reply by Joshua-Anderson** [reply]
Since F' has to support a wide variety of embedded systems toolchains, we try to be pretty conservative on what compiler features we require for the core framework. My fear would be excluding compilers with C++11 but not C++14 support. Going from C++03 to C++11 transitions wins us a lot of nice features and allows us to modernize the codebase. The benefits of C++14 are nice but don't seem valuable enough to drop support for older compilers.

**Reply by astroesteban** [reply]
Good points. Just going to C++11 itself gives the project huge benefits.

**Comment by SterlingPeet** [comment]
Last I checked C++ 14 didn’t compile correctly with GCC for my platform (ATmega AVRs), but C++11 does.  I’m not actively developing for that right now, but I might have a better feel for it in about 2 weeks.

Sterling

Sent from my mobile.

On Jun 26, 2021, at 11:02 AM, Esteban Duran ***@***.***> wrote:

﻿

Why not skip C++11 and jump to C++14? C++14 fixes several holes in C++11. It's basically a hotfix for 11.

Language Features

  *   C++11’s version of constexpr implies const for member functions, this is changed in C++14
  *   C++11 is missing auto return type deduction for regular functions (lambdas have it)
  *   C++11 does not have auto or variadic lambda parameters
  *   C++14 adds [[deprecated]] attribute
  *   C++14 adds ' digit separator, example: 1'000'000
  *   constexpr functions can be more than just a single return statement in C++14

Library Features

  *   std::make_unique was added in C++14, which enables the “no raw new” standard
  *   C++11 doesn’t have std::exchange
  *   C++14 adds some constexpr support for std::array cbegin, cend, crbegin and crend free functions added for consistency with begin and end free functions and member functions added to standard containers in C++11.”

**Reply by astroesteban** [reply]
Interesting. I wonder why…

**Reply by Joshua-Anderson** [reply]
Thanks Sterling, do you happen to know which GCC version you're using for the ATMega?

**Comment by SterlingPeet** [comment]
Not sure, but I will definitely check on this the next time I’m working on a relevant project.  Should be within a week or two.

Sterling

Sent from my mobile.

On Jun 26, 2021, at 2:28 PM, Esteban Duran ***@***.***> wrote:

﻿

Interesting. I wonder why…

**Comment by LeStarch** [comment, marked as answer]
It sounds like there are no active objections. I am going to mark this as answered.
